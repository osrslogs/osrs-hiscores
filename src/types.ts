import { AxiosProxyConfig } from 'axios';

export type Mode = 'normal' | 'ironman' | 'ultimate' | 'hardcore' | 'deadman' | 'seasonal' | 'tournament';

export type SkillName =
  | 'overall'
  | 'attack'
  | 'defence'
  | 'strength'
  | 'hitpoints'
  | 'ranged'
  | 'prayer'
  | 'magic'
  | 'cooking'
  | 'woodcutting'
  | 'fletching'
  | 'fishing'
  | 'firemaking'
  | 'crafting'
  | 'smithing'
  | 'mining'
  | 'herblore'
  | 'agility'
  | 'thieving'
  | 'slayer'
  | 'farming'
  | 'runecraft'
  | 'hunter'
  | 'construction';

export type BossName =
  | 'abyssalSire'
  | 'alchemicalHydra'
  | 'barrowsChests'
  | 'bryophyta'
  | 'callisto'
  | 'cerberus'
  | 'chambersOfXeric'
  | 'chambersOfXericChallengeMode'
  | 'chaosElemental'
  | 'chaosFanatic'
  | 'commanderZilyana'
  | 'corporealBeast'
  | 'crazyArchaeologist'
  | 'dagannothPrime'
  | 'dagannothRex'
  | 'dagannothSupreme'
  | 'derangedArchaeologist'
  | 'generalGraardor'
  | 'giantMole'
  | 'grotesqueGuardians'
  | 'hespori'
  | 'kalphiteQueen'
  | 'kingBlackDragon'
  | 'kraken'
  | 'kreeArra'
  | 'krilTsutsaroth'
  | 'mimic'
  | 'nightmare'
  | 'obor'
  | 'sarachnis'
  | 'scorpia'
  | 'skotizo'
  | 'theGauntlet'
  | 'theCorruptedGauntlet'
  | 'theatreOfBlood'
  | 'thermonuclearSmokeDevil'
  | 'tzKalZuk'
  | 'tzTokJad'
  | 'venenatis'
  | 'vetion'
  | 'vorkath'
  | 'wintertodt'
  | 'zalcano'
  | 'zulrah';

export type ClueType = 'all' | 'beginner' | 'easy' | 'medium' | 'hard' | 'elite' | 'master';

export type LeaguePoints = 'leaguePoints';

export type lastManStanding = 'lastManStanding';

export type BountyHunterType = 'hunter' | 'rogue';

export type ActivityName = LeaguePoints | BountyHunterType | ClueType | lastManStanding | BossName;

export type StatName = ActivityName | SkillName;

export interface Config {
  userAgent?: string;
  timeout?: number;
  proxy?: AxiosProxyConfig;
}

export interface RankedStat {
  rank: number;
}

export interface Activity extends RankedStat {
  rank: number;
  score: number;
}

export interface Skill extends RankedStat {
  rank: number;
  level: number;
  experience: number;
}

export interface CsvContent {
  skills: string[][];
  activities: string[][];
  remaining: string[][];
}

export interface PlayerSkillRow extends Skill {
  name: string;
  dead?: boolean; // only included on hc skill rows
}

export interface PlayerActivityRow extends Activity {
  name: string;
  dead?: boolean; // only included on hc activty rows
}

export type RankedStats = {
  [name in string]: RankedStat;
};

export type Skills = {
  [name in SkillName]: Skill;
};

export type Activities = {
  [name in ActivityName]: Activity;
};

export type Bosses = {
  [name in BossName]: Activity;
};

export type Clues = {
  [name in ClueType]: Activity;
};

export type BountyHunter = {
  [name in BountyHunterType]: Activity;
};

export interface Stats {
  skills: Skills;
  leaguePoints: Activity;
  bountyHunter: BountyHunter;
  clues: Clues;
  lastManStanding: Activity;
  bosses: Bosses;
}

export interface DisplayName {
  format: string;
}
