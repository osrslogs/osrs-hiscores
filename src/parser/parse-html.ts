import cheerio from 'cheerio';
import { PlayerSkillRow, PlayerActivityRow } from '../types';
import { InvalidHtmlError } from '../util/error';

/**
 * Extracts text from html element
 *
 * @param {CheerioElement} element
 *
 * @returns {string}
 */
const textFromElement = (element: CheerioElement | undefined): string => {
  const innerText = element?.firstChild.data;
  return innerText ? innerText.replace(/\uFFFd/g, '') : '';
};

/**
 * Extracts number from html element
 *
 * @param {CheerioElement} element
 *
 * @returns {number}
 */
const numberFromElement = (element: CheerioElement): number => {
  const innerText = element.firstChild.data;
  const number = innerText ? innerText.replace(/,/g, '') : '-1';
  return parseInt(number, 10);
};

/**
 * Parses a hiscore skill table page
 *
 * @param {string} skillPage HTML page of a OSRS hiscore table
 *
 * @returns {PlayerSkillRow[]}
 *
 * @throws {InvalidHtmlError} If html structure is invalid
 */
export const parseSkillPage = (skillPage: string, isHardcore = false): PlayerSkillRow[] => {
  const $ = cheerio.load(skillPage);
  const hiscoreRows = $('.personal-hiscores__row').toArray();

  const players: PlayerSkillRow[] = hiscoreRows.map(row => {
    const cells = row.children.filter(element => element.name === 'td');

    const [rankElement, nameCell, levelElement, experienceElement] = cells;
    const nameElement = nameCell.children.find(element => element.name === 'a');

    const stats = {
      name: textFromElement(nameElement),
      rank: numberFromElement(rankElement),
      level: numberFromElement(levelElement),
      experience: numberFromElement(experienceElement),
    };

    // we only care about death status if it is a hardcore table
    if (isHardcore) {
      const isDead = !!nameCell.children.find(element => element.name === 'img');
      return { ...stats, dead: isDead };
    }

    return stats;
  });

  return players;
};

/**
 * Parses a hiscore activity table page
 *
 * @param {string} activityPage HTML page of a OSRS hiscore table
 *
 * @returns {PlayerActivityRow[]}
 *
 * @throws {InvalidHtmlError} If html structure is invalid
 */
export const parseActivityPage = (activityPage: string, isHardcore = false): PlayerActivityRow[] => {
  const $ = cheerio.load(activityPage);
  const hiscoreRows = $('.personal-hiscores__row').toArray();

  const players: PlayerActivityRow[] = hiscoreRows.map(row => {
    const cells = row.children.filter(element => element.name === 'td');

    const [rankElement, nameCell, scoreElement] = cells;
    const nameElement = nameCell.children.find(element => element.name === 'a');

    const stats = {
      name: textFromElement(nameElement),
      rank: numberFromElement(rankElement),
      score: numberFromElement(scoreElement),
    };

    // we only care about death status if it is a hardcore table
    if (isHardcore) {
      const isDead = !!nameCell.children.find(element => element.name === 'img');
      return { ...stats, dead: isDead };
    }

    return stats;
  });

  return players;
};

/**
 * Parses a display name page to find the highlighted display name
 *
 * @param displayNamePage HTML page of a OSRS hiscore table
 *
 * @returns {string} The display name
 *
 * @throws {InvalidHtmlError} If display name is not found
 */
export const parseDisplayNamePage = (displayNamePage: string): string => {
  const $ = cheerio.load(displayNamePage);
  const displayName = $('[style="color:#AA0022;"]')[1]?.children[0]?.data;

  if (displayName && displayName.length > 0) {
    return displayName.replace(/\uFFFd/g, '');
  }

  throw new InvalidHtmlError('Could not find display name', displayNamePage);
};
