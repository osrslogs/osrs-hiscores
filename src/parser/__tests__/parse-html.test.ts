import { parseSkillPage, parseActivityPage, parseDisplayNamePage } from '../parse-html';
import htmlConfig from '../../util/test-util/html-mock.json';
import { PlayerSkillRow, PlayerActivityRow } from '../../types';

describe('parseSkillPage', () => {
  it('it returns PlayerSkillRow[] when given a valid normal page', () => {
    const players: PlayerSkillRow[] = parseSkillPage(htmlConfig.skillPage.normal);
    expect(players).toMatchSnapshot();
  });

  it('it parses `isDead` correctly for dead hardcore ironmen', () => {
    const players: PlayerSkillRow[] = parseSkillPage(htmlConfig.skillPage.hardcore, true);
    expect(players).toMatchSnapshot();
  });
});

describe('parseActivityPage', () => {
  it('it returns PlayerActivityRow[] when given a valid normal page', () => {
    const players: PlayerActivityRow[] = parseActivityPage(htmlConfig.activityPage.normal);
    expect(players).toMatchSnapshot();
  });

  it('it parses `isDead` correctly for dead hardcore ironmen', () => {
    const players: PlayerActivityRow[] = parseActivityPage(htmlConfig.activityPage.hardcore, true);
    expect(players).toMatchSnapshot();
  });
});

describe('parseDisplayNamePage', () => {
  it('returns display name when html contains highlighted name', () => {
    const displayName: string = parseDisplayNamePage(htmlConfig.displayNamePage.valid);
    expect(displayName).toMatchSnapshot();
  });

  it('throws Error when html does not contain highlighted name', () => {
    expect(() => parseDisplayNamePage(htmlConfig.displayNamePage.invalid)).toThrow(Error);
  });
});
