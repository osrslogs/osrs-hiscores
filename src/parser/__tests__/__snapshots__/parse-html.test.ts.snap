// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseActivityPage it parses \`isDead\` correctly for dead hardcore ironmen 1`] = `
Array [
  Object {
    "dead": false,
    "name": "Im Phone Btw",
    "rank": 2,
    "score": 279,
  },
  Object {
    "dead": false,
    "name": "Rear mirror",
    "rank": 3,
    "score": 275,
  },
  Object {
    "dead": true,
    "name": "BlameTheGame",
    "rank": 4,
    "score": 253,
  },
]
`;

exports[`parseActivityPage it returns PlayerActivityRow[] when given a valid normal page 1`] = `
Array [
  Object {
    "name": "Nex-Zeus",
    "rank": 8,
    "score": 34,
  },
  Object {
    "name": "Kaempen",
    "rank": 9,
    "score": 33,
  },
  Object {
    "name": "Lads sesh",
    "rank": 10,
    "score": 29,
  },
]
`;

exports[`parseDisplayNamePage returns display name when html contains highlighted name 1`] = `"BobRoss"`;

exports[`parseSkillPage it parses \`isDead\` correctly for dead hardcore ironmen 1`] = `
Array [
  Object {
    "dead": false,
    "experience": 1,
    "level": 2,
    "name": "5th hcim LUL",
    "rank": 3,
  },
  Object {
    "dead": true,
    "experience": 1,
    "level": 2,
    "name": "Nechs",
    "rank": 4,
  },
  Object {
    "dead": false,
    "experience": 1,
    "level": 2,
    "name": "Farmer McGee",
    "rank": 5,
  },
]
`;

exports[`parseSkillPage it returns PlayerSkillRow[] when given a valid normal page 1`] = `
Array [
  Object {
    "experience": 200,
    "level": 99,
    "name": "Sathon",
    "rank": 1,
  },
  Object {
    "experience": 200,
    "level": 99,
    "name": "Sathon",
    "rank": 1,
  },
  Object {
    "experience": 200,
    "level": 99,
    "name": "Sathon",
    "rank": 1,
  },
]
`;
